### PR TITLE
Support python requests keyword arguments when making remote calls

### PR DESCRIPTION
**Pull request by banjoh**
Since this library is heavily dependent on `python requests` for making remote calls, it is necessary to pass down parameters such as `timeout` and `header` to requests methods.

This PR is related to #148

### CODE REVIEW

File: simple_salesforce/api.py [hunk 1]
```diff
@@ -37,8 +38,60 @@ def _warn_request_deprecation():
     )
 
 
+class RequestMixin(object):
+    def __init__(self, session=None):
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: `MixIn`-objects generally do not initialize themselves. See examples here: https://github.com/pallets/werkzeug/blob/master/werkzeug/wrappers.py#L1281-L1628
https://github.com/django/django/blob/master/django/contrib/auth/mixins.py

Also this blog post explains some concepts: https://www.ianlewis.org/en/mixins-and-python

Basic idea behind `Mixins` (as I understand them), is that you give devs possibilities to add "automatic" features to other classes. In this case, your `Mixin` would not implement `_call_salesforce` method, it would implement something like `request`. The `Salesforce`-class would then always use `self.request()`-method but depending on the `Mixin` inherited, the behaviour would change.

I recommend either renaming this class, or thinking how to implement it in a more "generic" mixin style.

File: simple_salesforce/api.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 # has to be defined prior to login import
 DEFAULT_API_VERSION = '29.0'
+TIMEOUT = 60
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: I don't think we want to have `TIMEOUT` as a global property... If the idea is for the user of `Salesforce` to pass the `timeout` parameter for `requests`, `timeout` should be empty by default.

File: simple_salesforce/api.py [hunk 3]
```diff
@@ -37,8 +38,60 @@ def _warn_request_deprecation():
...
+    def __init__(self, session=None):
+        """
+        Mixin class to encapsulate all remote operations
+        * session -- Request's session object
+        """
+        self.session = session or requests.Session()
+        self.session_id = "sessionId"
+        self.name = ""
+
+    def _call_salesforce(self, method, url, **kwargs):
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: Previously you defined `timeout` as a global property, it could be also implemented here as an argument for this function. This would make more sense, since you can also provide default value for it.

File: simple_salesforce/api.py [hunk 4]
```diff
@@ -153,23 +206,19 @@ def __init__(
...
                          .format(instance=self.sf_instance,
                                  version=self.sf_version))
         self.apex_url = ('https://{instance}/services/apexrest/'
                          .format(instance=self.sf_instance))
 
-    def describe(self):
+    def describe(self, **kwargs):
         """Describes all available objects
+
+        Keyword Args: Parameters passed down to requests.request method
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: In this case `Keyword Args` is actually passed down to `requests.Sessions.request`-method, not `requests.request`. The documentation for those functions is (almost) the same, but we want to be accurate in our own documentation :).

Comment (CryptidVulpes) [inline simple_salesforce/api.py]: Remember the backticks for `requests.request`. The original documentation also uses `* kwargs` instead of `Keyword Args`. Is this change intentional?

File: simple_salesforce/api.py [hunk 5]
```diff
@@ -774,22 +807,12 @@ def __init__(self, username, password, security_token, sandbox=False,
 
 def _exception_handler(result, name=""):
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: This function has been defined as "private" (by Python naming standards). Why not remove it instead of marking it deprecated?

> _single_leading_underscore : weak "internal use" indicator. E.g. from M import * does not import objects whose name starts with an underscore.
-- https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces

> Use one leading underscore only for non-public methods and instance variables.
-- https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables

> Any backwards compatibility guarantees apply only to public interfaces. Accordingly, it is important that users be able to clearly distinguish between public and internal interfaces.
> 
> Documented interfaces are considered public, unless the documentation explicitly declares them to be provisional or internal interfaces exempt from the usual backwards compatibility guarantees. All undocumented interfaces should be assumed to be internal.
-- https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: simple_salesforce/login.py [hunk 6]
```diff
@@ -21,7 +21,7 @@
 def SalesforceLogin(
         username=None, password=None, security_token=None,
         organizationId=None, sandbox=False, sf_version=DEFAULT_API_VERSION,
-        proxies=None, session=None, client_id=None):
```
Comment (CryptidVulpes) [inline simple_salesforce/login.py]: Where did the documented `proxies` Keyword Argument vanish?

File: simple_salesforce/tests/test_api.py [hunk 7]
```diff
@@ -575,12 +577,14 @@ def setUp(self):
         self.mockresult = Mock()
         self.mockresult.url = 'http://www.example.com/'
         self.mockresult.json.return_value = 'Example Content'
+        self.mock_mixin = RequestMixin()
```
Comment (CryptidVulpes) [inline simple_salesforce/tests/test_api.py]: Why is this called `mock_mixin` when it is, in fact, the real `Mixin`-class?

File: simple_salesforce/tests/test_api.py [hunk 8]
```diff
@@ -630,8 +635,62 @@ def test_generic_error_code(self):
...
         self.assertEqual(str(cm.exception), (
             'Error Code 500. Response content'
             ': Example Content'))
+
+    def test_deprecated_exception_handler(self):
+        """Test calling deprecated exception handler"""
+        self.mockresult.status_code = 404
+        resource_name = 'SpecialContacts'
+        with self.assertRaises(SalesforceResourceNotFound) as exception:
+            _exception_handler(result=self.mockresult, name=resource_name)
```
Comment (CryptidVulpes) [inline simple_salesforce/tests/test_api.py]: If you remove the (currently)  deprecated `_exception_handler`, you can also remove this test and the import for `_exception_handler` :).

File: simple_salesforce/api.py (hunk 2 above)
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: It could also be configurable when you instantiate the `Salesforce` class...

File: simple_salesforce/api.py (hunk 5 above)
Comment (banjoh) [inline simple_salesforce/api.py]: That was my initial approach but I thought its possible to have someone (mis)use it since its not really private. I would rather go through the deprecation processes.

The project has been forked quite a bit. I would not want to break extensions that exist already.

File: simple_salesforce/login.py (hunk 6 above)
Comment (banjoh) [inline simple_salesforce/login.py]: It's one of the requests kwargs and hence part of the requests' documentation.

https://github.com/kennethreitz/requests/blob/master/requests/api.py#L16

File: simple_salesforce/api.py (hunk 5 above)
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: But as PEP8 describes, it is never a guarantee for undocumented private-functions to stay backward compatible. My personal opinion is to remove this function, but the upstream author has, of course, the ultimate say :).

File: simple_salesforce/api.py (hunk 1 above)
Comment (banjoh) [inline simple_salesforce/api.py]: Well, I think this is subject to different interpretations, but I see your point. I'll make some changes to simplify the implementation. Using a `mixin` may stir the wrong interpretation.

File: simple_salesforce/tests/test_api.py (hunk 8 above)
Comment (banjoh) [inline simple_salesforce/tests/test_api.py]: I added it there as a regression test. Once it's really deprecated then it can be removed.

Comment (banjoh) [inline simple_salesforce/tests/test_api.py]: I added it there as a regression test. Once it's really deprecated then it can be removed.

File: simple_salesforce/api.py (hunk 5 above)
Comment (banjoh) [inline simple_salesforce/api.py]: I'll probably remove it anyway. Let's see what the maintainer comments. I'll add a comment about this.

File: simple_salesforce/api.py [hunk 9]
```diff
@@ -38,7 +37,7 @@ def _warn_request_deprecation():
     )
 
 
-class RequestMixin(object):
+class RequestBase(object):
     def __init__(self, session=None):
         """
         Mixin class to encapsulate all remote operations
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: Missed a docstring here :)

File: simple_salesforce/api.py [hunk 10]
```diff
@@ -810,7 +809,7 @@ def _exception_handler(result, name=""):
     warnings.warn('This global exception handler is deprecated. It has been moved'
                   'to the RequestMixin class.', DeprecationWarning)
 
-    mixin = RequestMixin()
+    mixin = RequestBase()
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: `mixin` name appears here as well.

File: simple_salesforce/api.py [hunk 11]
```diff
@@ -61,7 +60,7 @@ def _call_salesforce(self, method, url, **kwargs):
         }
         additional_headers = kwargs.pop('headers', dict())
         headers.update(additional_headers or dict())
-        timeout = kwargs.pop('timeout', TIMEOUT)
+        timeout = kwargs.pop('timeout', 60)
```
Comment (CryptidVulpes) [inline simple_salesforce/api.py]: I still think you should add `timeout` as an argument with a default value, make it configurable in the `__init__()`-function or keep it unchanged in `kwargs`. This `timeout = kwargs.pop` looks too verbose...

File: simple_salesforce/tests/test_api.py [hunk 12]
```diff
@@ -577,8 +576,8 @@ def setUp(self):
         self.mockresult = Mock()
         self.mockresult.url = 'http://www.example.com/'
         self.mockresult.json.return_value = 'Example Content'
-        self.mock_mixin = RequestMixin()
-        self.exec_handler = self.mock_mixin._exception_handler
+        self.mixin = RequestBase()
```
Comment (CryptidVulpes) [inline simple_salesforce/tests/test_api.py]: `mixin` name here as well.
